hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,313 @@
+
+            #[cfg(any(
+                feature = "samd51",
+                feature = "same51",
+                feature = "same53",
+                feature = "same54"
+            ))]
+            d.chprilvl.modify(|_, w| w.prilvl().bits(lvl as u8));
+
+            if enable_interrupts {
+                // Enable all interrupt sources
+                d.chintenset
+                    .modify(|_, w| w.susp().set_bit().tcmpl().set_bit().terr().set_bit());
+            }
+        });
+
+        Channel { _status: Ready }
+    }
+
+    #[inline(always)]
+    fn _reset_private(&mut self, dmac: &mut DMAC) {
+        self.with_chid(dmac, |d| {
+            // Reset the channel to its startup state and wait for reset to complete
+            d.chctrla.modify(|_, w| w.swrst().set_bit());
+            while d.chctrla.read().swrst().bit_is_set() {}
+        })
+    }
+
+    #[inline(always)]
+    fn _trigger_private(&mut self, dmac: &mut DMAC) {
+        // SAFETY: This is safe because we are writing the correct channel
+        // number into the register
+        unsafe {
+            dmac.swtrigctrl.modify(|_, w| w.bits(1 << ID));
+        }
+    }
+}
+
+/// These methods may only be used on a `Ready` DMA channel
+impl<const ID: u8> Channel<Ready, ID> {
+    /// Issue a software reset to the channel. This will return the channel to
+    /// its startup state
+    pub fn reset(mut self, dmac: &mut DMAC) -> Channel<Uninitialized, ID> {
+        self._reset_private(dmac);
+
+        Channel {
+            _status: Uninitialized,
+        }
+    }
+
+    /// Set the FIFO threshold length. The channel will wait until it has
+    /// received the selected number of Beats before triggering the Burst
+    /// transfer, reducing the DMA transfer latency.
+    #[cfg(any(
+        feature = "samd51",
+        feature = "same51",
+        feature = "same53",
+        feature = "same54"
+    ))]
+    pub fn fifo_threshold(&mut self, threshold: FifoThreshold, dmac: &mut DmaController) {
+        self.with_chid(dmac.as_mut(), |d| {
+            d.chctrla.modify(|_, w| w.threshold().bits(threshold as u8));
+        })
+    }
+
+    /// Set burst length for the channel, in number of beats. A burst transfer
+    /// is an atomic, uninterruptible operation.
+    #[cfg(any(
+        feature = "samd51",
+        feature = "same51",
+        feature = "same53",
+        feature = "same54"
+    ))]
+    pub fn burst_length(&mut self, burst_length: BurstLength, dmac: &mut DmaController) {
+        self.with_chid(dmac.as_mut(), |d| {
+            d.chctrla
+                .modify(|_, w| w.burstlen().bits(burst_length as u8));
+        })
+    }
+
+    /// Start transfer on channel using the specified trigger source.
+    ///
+    /// # Return
+    ///
+    /// A `Channel` with a `Busy` status.
+    pub(crate) fn start(
+        mut self,
+        trig_src: TriggerSource,
+        trig_act: TriggerAction,
+        dmac: &mut DMAC,
```

Comment:
You might consider separating the configuration of the trigger source and action from the act of starting the transfer. You've already done that for the most part, but the trigger source and action are an exception. I would change that. I think it would also help to cut down on the latency from the point of calling `start` (or whatever calls `start`) to the point of the transfer actually starting, because you don't need to change the trigger settings before starting.

Also, again, I would move up the `dmac` argument.